# Bioinformatics and Statistical Methods for Analysis of Single-Cell RNA Sequencing Data

**Course Overview**: This one-day short course is designed to provide participants with practical skills in bioinformatics and statistical methods for the analysis of single-cell RNA sequencing (scRNA-seq) data. The course will cover key steps in scRNA-seq analysis using popular software packages, with a focus on data preprocessing, quality control, dimensionality reduction, clustering, and differential expression analysis.

**Instructors:**  
Dr. Bruce Corliss, Research Bioinformtician  
Dr. Allison Dickey, Computational Biologist  

**Target Audience**: Biologists, bioinformaticians, and researchers interested in single-cell RNA sequencing data analysis. 
<br/>  
  
**Format:** single day workshop, 8 hours run time, remote and on-site.
<br/>  
  
_Required Materials:_
Laptop with R, Rstudio, and Rtools installed.
<br/>
  
**Prerequisites:**
1. Basic understanding of molecular biology and genomics.
2. Basic familiarity with the R programming language.
3. Basic familiarity of github (downloading files etc.).
<br/>  

**Learning Materials**
1. Powerpoint presentations.
2. R and Python scripts and code snippets.
3. Example datasets for hands-on analysis.
4. Recommended reading materials and online resources.
<br/>
<br/>
  
## Course Outline:

**Module 0:** Course Preparation
1. Software installation and setup (instructions found [here]).
2. Download datasets (see datasets page).
3. Test scripts to verify functionality (run scripts found M0 folder).
  
**Module 1:** Introduction to Single-Cell RNA Sequencing and Software Setup
1. Overview of scRNA-seq technology and its applications
2. Introduction to software tools: Cell Ranger, Seurat, SingleR, and Monocle

<br/>

**Module 2:** Data Preprocessing and Quality Control using Cell Ranger
1. Preprocessing steps using Cell Ranger.
2. Quality control metrics and filtering strategies.
3. Data exploration with Cell Ranger outputs.
<br/>

**Module 3:** Data processing using Seurat
1. Introduction to Seurat: an R toolkit for single-cell genomics.
2. Data normalization and structure of S4 class Seurat object.
3. Horizontal and vertical integration of scRNA datasets.
4. Linear and non-linear dimension reductions.
<br/>  

**Module 4:** Exploratory Data Analysis with Seurat (R)
1. Dimensionality reduction techniques (PCA, t-SNE, UMAP).
2. Visualization of scRNA-seq data using Seurat.
<br/>

**Module 5:** Cell Type Identification and Clustering with Seurat
1. Marker based cell identification using Seurat and SingleR.
2. Atlast based cell identification.
<br/>

**Module 6:** Differential Expression Analysis with Seurat
1. Statistical methods for identifying differentially expressed genes using Seurat.
2. Visualizing and interpreting differential expression results.
3. Accounting for batch effects in differential expression analysis.
<br/>

**Module 7:** Best Practices and Troubleshooting
1. Best practices for scRNA-seq analysis.
2. Troubleshooting and common challenges.
3. FAIR principles in data science research.
<br/>

**Module 8:** Q&A and Wrap-Up
1. Open discussion and Q&A session.
2. Summary of key concepts and take-home messages.
<br/>



_Evaluation:_ Participants will be evaluated based on their engagement in the hands-on sessions, participation in discussions, and completion of assigned exercises.
<br/>

**Certificate of Completion:** Participants who successfully complete the course will receive a certificate of completion.
